src/librbd/api/Migrate.cc

```diff
@@ -0,0 +1,991 @@
+    if (r < 0) {
+      lderr(m_cct) << "failed retrieving snapshot status: " << cpp_strerror(r)
+                   << dendl;
+      return r;
+    }
+    if (is_protected) {
+      lderr(m_cct) << "image has protected snapshot '" << snap.name << "'"
+                   << dendl;
+      return -EBUSY;
+    }
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::set_migration() {
+  if (m_src_old_format) {
+    return v1_set_migration();
+  } else {
+    return v2_set_migration();
+  }
+}
+
+template <typename I>
+int Migrate<I>::v1_set_migration() {
+  ldout(m_cct, 10) << dendl;
+
+  bufferlist header;
+  int r = read_header_bl(m_src_io_ctx, m_src_header_oid, header, nullptr);
+  if (r < 0) {
+    lderr(m_cct) << "failed reading " << m_src_header_oid << " header: "
+                 << cpp_strerror(r) << dendl;
+    return r;
+  }
+  if (header.length() < sizeof(rbd_obj_header_ondisk)) {
+    lderr(m_cct) << "v1 header too small" << dendl;
+    return -EIO;
+  }
+  if (memcmp(RBD_HEADER_TEXT, header.c_str(), sizeof(RBD_HEADER_TEXT)) != 0) {
+    lderr(m_cct) << "unrecognized v1 header" << dendl;
+    return -ENXIO;
+  }
+
+  r = cls_client::migration_set(&m_src_io_ctx, m_src_header_oid,
+                                m_src_migration_spec);
+  if (r < 0) {
+    lderr(m_cct) << "failed to set migration header: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  bufferlist::iterator it = header.begin();
+  static_assert(sizeof(RBD_HEADER_TEXT) == sizeof(RBD_MIGRATE_HEADER_TEXT),
+                "length of rbd headers must be the same");
+  it.copy_in(sizeof(RBD_MIGRATE_HEADER_TEXT), RBD_MIGRATE_HEADER_TEXT);
+
+  r = m_src_io_ctx.write(m_src_header_oid, header, header.length(), 0);
+  if (r < 0) {
+    lderr(m_cct) << "failed writing " << m_src_header_oid
+                 << " header: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  r = tmap_rm(m_src_io_ctx, m_src_image_name);
+  if (r < 0) {
+    lderr(m_cct) << "failed removing " << m_src_image_name << " from tmap: "
+                 << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::v2_set_migration() {
+  ldout(m_cct, 10) << dendl;
+
+  v2_unlink_image();
+
+  int r = cls_client::migration_set(&m_src_io_ctx, m_src_header_oid,
+                                    m_src_migration_spec);
+  if (r < 0) {
+    lderr(m_cct) << "failed to set migration header: " << cpp_strerror(r)
+                 << dendl;
+    if (r == -EOPNOTSUPP) {
+      v2_relink_image();
+    }
+    return r;
+  }
+
+  r = m_image_ctx->operations->update_features(RBD_FEATURE_MIGRATING, true);
+  if (r < 0) {
+    lderr(m_cct) << "failed enabling migrating feature: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::unset_migration() {
+  ldout(m_cct, 10) << dendl;
+
+  int r;
+
+  if (m_image_ctx->old_format) {
```

Comment:
... since ```m_image_ctx``` is now referencing the destination image, it's odd that this could be ```true```

Reply:
`unset_migration()` is reused by `abort()`, and in that case `m_image_ctx` is referencing the source image.